internal/provisioner/cluster.go

```diff
@@ -234,6 +258,38 @@ func DeleteCluster(clusterID, s3StateStore string, logger log.FieldLogger) error
 	return nil
 }
 
+// waitForKubernetesReadyness will poll a given kubernetes cluster at a regular
+// interval for it to become ready. If the cluster fails to become ready before
+// the provided timeout then an error will be returned.
+func waitForKubernetesReadyness(dns string, kops *kops.Cmd, timeout time.Duration) error {
+	ready := make(chan bool, 1)
+	defer close(ready)
+
+	go func() {
+		for {
```

Comment:
This goroutine will never terminate if the validation never succeeds -- in practice, the CLI will exit so we're probably fine in this implementation, but this could leak resources in a long-running server.

I'm guessing we aren't actually worried about `kops validate` hanging? (If so, the following is invalid.)

Might we just eliminate the goroutine altogether and collapse this function to something like:
```go
func waitForKubernetesReadiness(dns string, kops *kops.Cmd, timeout time.Duration) error {
    timer := time.NewTimer(timeout * time.Second)
    defer timer.Stop()

    for {
        select {
        case <-timer.C:
            return errors.New("timed out waiting for k8s cluster to become ready")
        default:
            err := kops.ValidateCluster(dns, true)
            if err == nil {
                return nil
            }
            time.Sleep(5 * time.Second)
        }
    }

    return nil
}
```

Reply:
I was sort of worried about `kops validate` hanging, but that's probably not worth the extra complexity here. This solution seems like it may be the better first implementation and we can tackle strange kops-hanging issues should we run into them.